core/s3.py

```diff
@@ -521,6 +541,24 @@ def marc_file_url(self, library, lane, end_time, start_time=None):
         parts = [time_part, lane.display_name]
         return root + self.key_join(parts) + ".mrc"
 
+    def analytics_file_url(
+        self,
+        library: Library,
+        license_pool: LicensePool,
+        end_time: datetime,
+        start_time: datetime = None,
+    ):
+        """The path to the analytics data file for the given library, license
+        pool and date range."""
+        bucket = self.get_bucket(S3UploaderConfiguration.ANALYTICS_BUCKET_KEY)
+        root = self._analytics_file_root(bucket, library)
+        if start_time:
+            time_part = str(start_time) + "-" + str(end_time)
+        else:
+            time_part = str(end_time)
+        parts = [time_part, license_pool.collection_id]
+        return root + self.key_join(parts) + ".json"
```

Comment:
Although unlikely, it is possible that we can have overlapping times for events on the same collection, thus causing a loss in data through inadvertent overwrites.
We can add some random text at the end of each file name to better ensure uniqueness eg. `16-1655277288.278079-thJIu786.json`
Overlaps become more probable if we are directly using times from external event sources (like checkout events) that may not provide millisecond precision